common/quotas/limiter_test.go

```diff
@@ -88,6 +88,65 @@ func TestMultiStageRateLimitingMultipleDomains(t *testing.T) {
 	check(" after refill")
 }
 
+func TestMultiStageRateLimitingWaitDomainRPS(t *testing.T) {
+	t.Parallel()
+	policy := newFixedRpsMultiStageRateLimiter(t, 2, 1)
+
+	check := func(suffix string) {
+		withTimeout100(func(ctx context.Context) {
+			assert.NoError(t, policy.Wait(ctx, Info{Domain: "one"}), "first should work"+suffix)
+		})
+		withTimeout100(func(ctx context.Context) {
+			err := policy.Wait(ctx, Info{Domain: "one"})
+			assert.Error(t, err, "second should be limited"+suffix) // per domain limited
+			assert.ErrorIs(t, err, clock.ErrCannotWait, "should return ErrCannotWait when rate limited"+suffix)
+		})
+		withTimeout1010(func(ctx context.Context) {
+			assert.NoError(t, policy.Wait(ctx, Info{Domain: "one"}), "third should work after waiting 1.01s"+suffix) // domain token replenished
+		})
+	}
+
+	check("")
+	// allow bucket to refill
+	time.Sleep(time.Second)
```

Comment:
These tests are clever and demonstrate the behavior, but relying on a real clock both slows them down considerably (they'd run in <1 ms otherwise) and in some scenarios can introduce flakiness when run on a particularly slow machine. We've been trying to slowly remove any tests that use the real clock or need sleeping to pass.

I'd recommend either using a mock clock and manually advancing the time, or using policies that will always fail/succeed. Then there are just 4 cases to consider with the (globalLimiter, domainLimiter) as ((ok,ok), (ok,err), (err, ok), (err,err)). You could consider a table test with the policy, similar to the PR you made for task_list_limiter.